File: src/native/ping_pong_lb_caf/ping_pong_lb.cpp
/*
 * ping_pong_lb.cpp
 *
 * This source file implements a distributed message-passing system using the
 * Actor model in the CAF framework. It simulates clients sending messages
 * (pings) to servers through a load balancer. Servers process messages and
 * respond with pongs. The main process ensures orderly termination.
 *
 * Parameters:
 * - servers: Number of server processes.
 * - clients: Number of client processes.
 * - messages: Messages each client sends.
 * - verbose: Verbose mode for debugging.
 *
 * The system terminates when the clients, the servers and the load balancer
 * finish processing.
 */

#include <caf/actor_system.hpp>
#include <caf/all.hpp>
#include <caf/atom.hpp>
#include <caf/send.hpp>
#include <caf/stateful_actor.hpp>
#include <cstddef>
#include <iostream>
#include <vector>

using namespace caf;

// Server behavior: Processes "ping" messages from clients and responds with
// "pong". The server notifies the main actor upon termination.
behavior server_fun(event_based_actor *self, const actor &main,
                    const bool verbose) {
  return {[=](const actor &client, atom_value msg) {
            if (msg == atom("ping")) {
              if (verbose)
                aout(self) << "Server <" << self->id()
                           << "> received: " << to_string(msg) << std::endl;
              self->send(client, atom("pong"));
            }
          },
          [=](atom_value msg) {
            if (msg == atom("stop")) {
              if (verbose)
                aout(self) << "Server <" << self->id() << "> terminating."
                           << std::endl;
              self->send(main, atom("server"), self->id(), atom("done"));
              self->quit();
            }
          }};
}

// Load balancer state to store the index of the current server
struct load_balancer_state {
  size_t server_index = 0;
};

// Load balancer: Distributes the messages to servers in a round-robin manner.
// The load balancer notifies the main actor upon termination.
behavior load_balancer_fun(stateful_actor<load_balancer_state> *self,
                           const std::vector<actor> &servers, const actor &main,
                           const bool verbose) {

  return {[=](const actor &client, atom_value ping) mutable {
            if (ping == atom("ping")) {
              self->send(servers[self->state.server_index], client, ping);
              self->state.server_index =
                  (self->state.server_index + 1) % servers.size();
            }
          },
          [=](atom_value msg) mutable {
            if (msg == atom("stop")) {
              if (verbose)
                aout(self) << "Load balancer <" << self->id()
                           << "> terminating." << std::endl;
              self->send(main, atom("lb"), self->id(), atom("done"));
              self->quit();
            }
          }};
}

// Client behavior: Sends a predefined number of "ping" messages to the
// load balancer and waits for "pong" responses. It notifies the main actor upon
// completion.
behavior client_fun(event_based_actor *self, const actor &load_balancer,
                    const size_t num_messages, const actor &main,
                    const bool verbose) {
  size_t responses_received = 0;

  return {[=](atom_value msg) mutable {
    if (msg == atom("start")) {
      self->send(load_balancer, self, atom("ping"));
    }
    if (msg == atom("pong")) {
      ++responses_received;
      if (verbose)
        aout(self) << "Client <" << self->id()
                   << "> received: " << to_string(msg) << std::endl;
      if (responses_received == num_messages) {
        if (verbose)
          aout(self) << "Client <" << self->id()
                     << "> reached max pings, terminating." << std::endl;
        self->send(main, atom("client"), self->id(), atom("done"));
        self->quit();
      } else {
        self->send(load_balancer, self, atom("ping"));
      }
    }
  }};
}

// Main actor state to store the server actors, the load balancer and the number
// of finished servers and clients
struct main_state {
  actor load_balancer;
  std::vector<actor> servers;
  size_t servers_finished = 0;
  size_t clients_finished = 0;
};

// Main actor: Waits for all clients and the server to finish before terminating
// the system.
behavior main_fun(stateful_actor<main_state> *self, const size_t num_clients,
                  const bool verbose) {

  return {
      [=](const actor &load_balancer) mutable {
        self->state.load_balancer = load_balancer;
      },

      [=](const std::vector<actor> servers) mutable {
        self->state.servers = servers;
      },

      [=](atom_value type, actor_id id, atom_value status) mutable {
        if (status == atom("done")) {
          if (type == atom("client")) {
            ++self->state.clients_finished;
            if (verbose) {
              aout(self) << "Client <" << id << "> finished." << std::endl;
            }
            if (self->state.clients_finished == num_clients) {
              aout(self) << "All clients finished, terminating servers."
                         << std::endl;
              for (auto &server : self->state.servers) {
                self->send(server, atom("stop"));
              }
            }
          } else if (type == atom("server")) {
            ++self->state.servers_finished;
            if (verbose) {
              aout(self) << "Server <" << id << "> finished." << std::endl;
            }
            if (self->state.servers_finished == self->state.servers.size()) {
              aout(self) << "All servers finished, terminating load balancer."
                         << std::endl;
              self->send(self->state.load_balancer, atom("stop"));
            }
          } else if (type == atom("lb")) {
            if (verbose) {
              aout(self) << "Load balancer <" << id << "> finished."
                         << std::endl;
            }
            aout(self) << "Terminating." << std::endl;
            self->quit();
          }
        }
      }};
}

// Configuration structure for setting up the actor system parameters
struct config : public actor_system_config {
  size_t num_servers = 2;
  size_t num_clients = 8;
  size_t num_messages = 10;
  bool verbose = false;

  config() {
    opt_group{custom_options_, "global"}
        .add(num_servers, "servers", "Number of Servers")
        .add(num_clients, "clients", "Number of Clients")
        .add(num_messages, "messages", "Number of Messages")
        .add(verbose, "verbose", "Verbose output");
  }
};

// Main function that initializes the actor system and spawns actors
void caf_main(actor_system &sys, const config &cfg) {
  // Spawn main actor to oversee termination
  auto main_actor = sys.spawn(main_fun, cfg.num_clients, cfg.verbose);

  // Spawn server actors
  std::cout << "Starting " << cfg.num_servers << " servers..." << std::endl;
  std::vector<actor> server_actors;
  for (size_t i = 0; i < cfg.num_servers; ++i) {
    server_actors.push_back(sys.spawn(server_fun, main_actor, cfg.verbose));
  }
  anon_send(main_actor, server_actors);

  // Spawn load balancer actor
  std::cout << "Starting load balancer..." << std::endl;
  auto load_balancer_actor =
      sys.spawn(load_balancer_fun, server_actors, main_actor, cfg.verbose);
  anon_send(main_actor, load_balancer_actor);

  // Spawn client actors
  std::cout << "Starting " << cfg.num_clients << " clients to send "
            << cfg.num_messages << " messages each..." << std::endl;
  for (size_t i = 0; i < cfg.num_clients; ++i) {
    auto client = sys.spawn(client_fun, load_balancer_actor, cfg.num_messages,
                            main_actor, cfg.verbose);
    anon_send(client, atom("start"));
  }
}

// Entry point for the CAF-based actor system
CAF_MAIN()

File: src/native/ping_pong_lb_erl/ping_pong_lb.erl
%% ping_pong_lb.erl
%%
%% This module implements a distributed message-passing system using the Actor model in Erlang.
%% It simulates clients sending messages (pings) to servers through a load balancer.
%% Servers process messages and respond with pongs. The main process ensures orderly termination.
%%
%% Parameters:
%% - NumServers: Number of server processes.
%% - NumClients: Number of client processes.
%% - NumMessages: Number of messages each client sends.
%% - verbose: Verbose mode for debugging.
%%
%% The system terminates when the clients, the servers and the load balancer finish processing.

-module(ping_pong_lb).
-export([main/0, main/1]).

% Entry point of the program
main() ->
    start(2, 8, 10, false).

main(Args) ->
    case parse_args(Args) of
        {ok, NumServers, NumClients, NumMessages, Verbose} ->
            start(NumServers, NumClients, NumMessages, Verbose); % Start the system with parsed parameters
        {error, Msg} ->
            io:format("Error: ~s~nUsage: main <num_servers> <num_clients> <num_messages> [verbose]~n", [Msg])
    end.

% Initializes the system: spawns servers, clients and load balancer
start(NumServers, NumClients, NumMessages, Verbose) ->
    Main = self(), % Store reference to the main process

    % Spawn server processes
    io:format("Starting ~p servers...~n", [NumServers]),
    Servers = [spawn(fun() -> server(Main, Verbose) end) || _ <- lists:seq(1, NumServers)],

    % Spawn load balancer process
    io:format("Starting load balancer...~n"),
    LoadBalancer = spawn(fun() -> load_balancer(Servers, 0, Main, Verbose) end),

    % Spawn client processes
    io:format("Starting ~p clients to send ~p messages each...~n", [NumClients, NumMessages]),
    [spawn(fun() -> client(LoadBalancer, NumMessages, Main, Verbose) end) || _ <- lists:seq(1, NumClients)],

    % Wait for all clients, servers and the load balancer to finish
    wait_for_processes(LoadBalancer, Servers, NumServers, NumClients, Verbose),

    io:format("Terminating.~n"),
    halt(0). % Shutdown the system

% Parses command-line arguments and converts them
parse_args([NumServersStr, NumClientsStr, NumMessagesStr | Rest]) ->
    case {string:to_integer(NumServersStr), string:to_integer(NumClientsStr), string:to_integer(NumMessagesStr)} of
        {{IntServer, _}, {IntClients, _}, {IntMessages, _}} ->
            Verbose = lists:member("verbose", Rest), % Check for verbosity flag
            {ok, IntServer, IntClients, IntMessages, Verbose};
        _ ->
            {error, "Invalid arguments."}
    end;
parse_args(_) ->
    {error, "Invalid number of arguments."}.


% Client process: Sends messages to the load balancer and waits for responses
client(_, 0, Main, true) ->
    io:format("Client ~p reached max pings, terminating.~n", [self()]),
    Main ! {client, self(), done}; % Notify the main process that client is done
client(_, 0, Main, false) ->
    Main ! {client, self(), done}; % Notify the main process that client is done
client(LoadBalancer, NumMessages, Main, true) -> % Verbose mode
    % Send a ping to the server
    LoadBalancer ! {self(), ping},

    % Wait for a pong message
    receive
        pong ->
            io:format("Client ~p received: pong~n", [self()]),
            client(LoadBalancer, NumMessages - 1, Main, true) % Recursively send more messages
    end;
client(LoadBalancer, NumMessages, Main, false) -> % Silent mode
    % Send a ping to the server
    LoadBalancer ! {self(), ping},

    % Wait for a pong message
    receive
        pong ->
            client(LoadBalancer, NumMessages - 1, Main, false) % Recursively send more messages
    end.

% Server process: Handles ping messages and responds with pong messages
server(Main, true) -> % Verbose mode enabled
    receive
        {From, ping} ->
            io:format("Server ~p received: ping~n", [self()]),
            From ! pong, % Respond with pong
            server(Main, true); % Recursively wait for more messages
        stop ->
            io:format("Server ~p terminating.~n", [self()]),
            Main ! {server, self(), done} % Notify the main process
    end;
server(Main, false) -> % Silent mode
    receive
        {From, ping} ->
            From ! pong, % Respond with pong
            server(Main, false); % Recursively wait for more messages
        stop ->
            Main ! {server, self(), done} % Notify the main process
    end.

% Load balancer: Distributes messages among servers in a round-robin manner
load_balancer(Servers, Index, Main, true) ->
    receive
        {From, Msg} ->
            % Select a worker using round-robin scheduling
            Server = lists:nth((Index rem length(Servers)) + 1, Servers),
            Server ! {From, Msg}, % Forward message to worker
            load_balancer(Servers, (Index + 1), Main, true); % Continue loop
        stop ->
            io:format("Load balancer ~p terminating.~n", [self()]),
            Main ! {lb, self(), done} % Notify the main process
    end;
load_balancer(Servers, Index, Main, false) ->
    receive
        {From, Msg} ->
            % Select a worker using round-robin scheduling
            Server = lists:nth((Index rem length(Servers)) + 1, Servers),
            Server ! {From, Msg}, % Forward message to worker
            load_balancer(Servers, (Index + 1), Main, false); % Continue loop
        stop ->
            Main ! {lb, self(), done} % Notify the main process
    end.

% Waiting loop for all clients to signal completion
wait_for_processes(LoadBalancer, _, 0, (-1), true) -> % Terminate load balancer when servers are done
    io:format("All servers finished, terminating load balancer.~n"),
    LoadBalancer ! stop,
    receive
        {lb, Pid, done} ->
            io:format("Load balancer ~p finished.~n", [Pid])
    end;
wait_for_processes(LoadBalancer, _, 0, (-1), false) -> % Terminate load balancer when servers are done
    io:format("All servers finished, terminating load balancer.~n"),
    LoadBalancer ! stop,
    receive
        {lb, _Pid, done} -> done
    end;
wait_for_processes(LoadBalancer, Servers, NumServers, (-1), true) ->
    receive
        {server, Pid, done} ->
            io:format("Server ~p finished.~n", [Pid]),
            wait_for_processes(LoadBalancer, Servers, NumServers - 1, (-1), true) % Decrement server count
    end;
wait_for_processes(LoadBalancer, Servers, NumServers, (-1), false) ->
    receive
        {server, _Pid, done} ->
            wait_for_processes(LoadBalancer, Servers, NumServers - 1, (-1), false) % Decrement server count
    end;
wait_for_processes(LoadBalancer, Servers, NumServers, 0, Verbose) -> % Terminate servers when clients are done
    io:format("All clients finished, terminating servers.~n"),
    lists:foreach(fun(Server) -> Server ! stop end, Servers), % Stop servers
    wait_for_processes(LoadBalancer, Servers, NumServers, (-1), Verbose);
wait_for_processes(LoadBalancer, Servers, NumServers, NumClients, true) ->
    receive
        {client, Pid, done} ->
            io:format("Client ~p finished.~n", [Pid]),
            wait_for_processes(LoadBalancer, Servers, NumServers, NumClients - 1, true) % Decrement client count
    end;
wait_for_processes(LoadBalancer, Servers, NumServers, NumClients, false) ->
    receive
        {client, _Pid, done} ->
            wait_for_processes(LoadBalancer, Servers, NumServers, NumClients - 1, false) % Decrement client count
    end.

